prov/util/src/util_attr.c

```diff
@@ -426,43 +426,76 @@ static int fi_resource_mgmt_level(enum fi_resource_mgmt rm_model)
 /* If a provider supports basic registration mode it should set the FI_MR_BASIC
  * mode bit in prov_mode. Support for FI_MR_SCALABLE is indicated by not setting
  * any of OFI_MR_BASIC_MAP bits. */
-int ofi_check_mr_mode(uint32_t api_version, uint32_t prov_mode,
+int ofi_check_mr_mode(const struct fi_provider *prov, uint32_t api_version,
+		      uint64_t user_info_caps, uint32_t prov_mode,
 		      uint32_t user_mode)
 {
+	uint64_t prov_mode_log;
+	int ret;
+
 	if (FI_VERSION_LT(api_version, FI_VERSION(1, 5))) {
+		if (!(user_info_caps & (FI_RMA | FI_ATOMIC)) &&
+		    !(prov_mode & FI_MR_LOCAL))
+			return 0;
+
 		prov_mode &= ~FI_MR_LOCAL; /* ignore local bit */
 
 		switch (user_mode) {
 		case FI_MR_UNSPEC:
-			return OFI_CHECK_MR_SCALABLE(prov_mode) ||
+			ret = OFI_CHECK_MR_SCALABLE(prov_mode) ||
 				OFI_CHECK_MR_BASIC(prov_mode) ?
 				0 : -FI_ENODATA;
+			break;
 		case FI_MR_BASIC:
-			return OFI_CHECK_MR_BASIC(prov_mode) ? 0 : -FI_ENODATA;
+			ret = OFI_CHECK_MR_BASIC(prov_mode) ? 0 : -FI_ENODATA;
+			break;
 		case FI_MR_SCALABLE:
-			return OFI_CHECK_MR_SCALABLE(prov_mode) ? 0 : -FI_ENODATA;
+			ret = OFI_CHECK_MR_SCALABLE(prov_mode) ? 0 : -FI_ENODATA;
+			break;
 		default:
-			return -FI_ENODATA;
+			ret = -FI_ENODATA;
+			break;
 		}
 	} else {
+		if (!(user_info_caps & (FI_RMA | FI_ATOMIC))) {
+			if (prov_mode & FI_MR_LOCAL)
+				prov_mode &= ~(FI_MR_PROV_KEY | FI_MR_VIRT_ADDR);
+			else
+				prov_mode &= ~OFI_MR_BASIC_MAP;
```

Comment:
Maybe re-structuring this to:
```
prov_mode &= ~(FI_MR_PROV_KEY | FI_MR_VIRT_ADDR);
if (!(prov_mode & FI_MR_LOCAL))
    prov_mode &= ~FI_MR_ALLOCATED;
```
would help clarify the intent behind the changes?

Reply:
👍